oeps/oep-0012.rst

```diff
@@ -0,0 +1,520 @@
+new components will be implemented using a new Django view subclass called
+`DjangoComponentView`. These views are rendered server-side, and an API is
+provided to allow clients to dynamically render a component if required.
+
+Motivation
+==========
+
+There are a number of driving use cases that act as motivation for supporting
+pluggable user interfaces.
+
+Use Case 1: LMS Course Tabs
+---------------------------
+
+A very common need for new features is to be able to add a new course tab for
+the learner, for example Discussions, Student Notes, and Teams. There is a
+supported pluggable entry point for course tabs (see `Adding a new course tab`_)
+but this requires that the feature add a new URL for the feature, and render the
+whole page at that URL using a Mako template. It would be preferable for a
+course tab to be able to register a pluggable user interface that will be
+presented as a tab, and for the implementor not to need to know how to implement
+the full page.
+
+There has been a long standing request to support rendering XBlocks as course
+tabs (`Allow xblocks to be shown as course tabs`_), but there are conceptual
+issues with this approach:
+
+* XBlocks must be configured and saved into the modulestore which doesn't make
+  much sense for course tabs that don't have their own state
+
+* Replacing a built-in tab such as "Course" would require a data migration for
+  every course if it was saved as an XBlock
+
+* It seems confusing to make an XBlock that should never be added into the
+  courseware, but which only exists to be shown as a tab
+
+It would be prefable to allow a feature library to register a course tab plugin
+which in turn renders using a component. This allows the feature to provide a
+custom UI without having to add custom blocks to each course, and if the feature
+is removed then the tab will seamlessly disappear without having to remove the
+blocks again.
+
+Use Case 2: LMS Discussions
+---------------------------
+
+Open edX Discussions is a complex feature that integrates into many parts of the
+LMS (currently as an inline XBlock, as a courseware tab, and to provide team
+discussions). Ideally, Open edx Discussions would be a separately-installable
+feature that would provide plugins for all of this functionality, and then an
+Open edX operator could choose whether to install edX discussions or an
+alternate collaboration tool.
+
+Currently the discussion board is a Django view built into edx-platform, but it
+should instead be made into a plugin. One option would be to create a new
+CourseDiscussionXBlock to display the "Course" tab, but as described above there
+are downsides to forcing this to be an XBlock. It would be cleaner if the edX
+discussion library provided a pluggable user interface that can be dropped into
+each of the contexts in which it is needed. The component would be provided
+with the context for authorization and to customize itself appropriately.
```

Comment:
That's exactly what an XBlock is intended to be -- a pluggable user interface which can be plugged in where ever needed.